### PR TITLE
Wasm to x86 backend

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
towards #1222

### CODE REVIEW

File: src/libasr/codegen/wasm_to_x86.cpp [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+    CodeGenError(const std::string &msg, const Location &loc)
+        : d{diag::Diagnostic(msg, diag::Level::Error, diag::Stage::CodeGen,
+                             {diag::Label("", {loc})})} {}
+};
+
+}  // namespace
+
+namespace wasm {
+
+class WASMDecoder {
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/wasm_to_x86.cpp:43]: This `WASMDecoder` class is being repeated (it is already defined in `src/libasr/codegen/wasm_to_wat.cpp`). We should extract it and place in a common file.

**doubt:** Where should we place this `WASMDecoder` class? Creating a new file (e.g. `wasm_disassembler.cpp` or `wasm_decoder.cpp`) seems possible, but I am unsure if this is the right way.

**Comment by certik** [comment]
Awesome, thanks for this! Will review it tomorrow.

File: src/libasr/codegen/wasm_to_x86.cpp (hunk 1 above)
Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_to_x86.cpp]: I temporarily moved this class into a new file `wasm_decoder.h`.

**Comment by Shaikh-Ubaid** [comment]
I am an facing the following error. 
```bash
Traceback (most recent call last):
  Binary file "/home/ubaid/OpenSource/lpython/src/bin/lpython", in _start()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __libc_start_main()
  File "/home/ubaid/OpenSource/lpython/src/bin/lpython.cpp", line 1555, in ??
    return compile_to_binary_wasm_to_x86(arg_file, outfile,
  File "/home/ubaid/OpenSource/lpython/src/bin/lpython.cpp", line 856, in ??
    LFortran::Result<int> res = LFortran::wasm_to_x86(r3.result, al,  outfile, time_report, diagnostics);
  File "/home/ubaid/OpenSource/lpython/src/libasr/codegen/wasm_to_x86.cpp", line 98, in LFortran::wasm_to_x86(LFortran::Vec<unsigned char>&, Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LFortran::diag::Diagnostics&)
    x86_generator.get_x86_bytes(m_a);
  File "/home/ubaid/OpenSource/lpython/src/libasr/codegen/wasm_to_x86.cpp", line 53, in LFortran::wasm::X86Generator::get_x86_bytes(LFortran::X86Assembler&)
    v.decode_instructions();
  File "/home/ubaid/OpenSource/lpython/src/libasr/../libasr/wasm_visitor.h", line 926, in LFortran::WASM_INSTS_VISITOR::BaseWASMVisitor<LFortran::WASM_INSTS_VISITOR::X86Visitor>::decode_instructions()
    self().visit_Call(funcidx);
  File "/home/ubaid/OpenSource/lpython/src/libasr/../libasr/codegen/wasm_to_x86.h", line 52, in LFortran::WASM_INSTS_VISITOR::X86Visitor::visit_Call(unsigned int)
    m_a.asm_call_label(exports[func_index].name);
  File "/home/ubaid/OpenSource/lpython/src/libasr/../libasr/codegen/x86_assembler.h", line 691, in LFortran::X86Assembler::asm_call_label(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    uint32_t imm32 = relative_symbol(label);
  File "/home/ubaid/OpenSource/lpython/src/libasr/../libasr/codegen/x86_assembler.h", line 365, in LFortran::X86Assembler::relative_symbol(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    return reference_symbol(name, 2).value-pos()-4;
  File "/home/ubaid/OpenSource/lpython/src/libasr/../libasr/codegen/x86_assembler.h", line 333, in LFortran::X86Assembler::reference_symbol(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)
    if (m_symbols.find(name) == m_symbols.end()) {
  File "/usr/include/c++/9/bits/stl_map.h", line 1169, in std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LFortran::Symbol, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol> > >::find(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    { return _M_t.find(__x); }
  File "/usr/include/c++/9/bits/stl_tree.h", line 2562, in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol> > >::find(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    iterator __j = _M_lower_bound(_M_begin(), _M_end(), __k);
  File "/usr/include/c++/9/bits/stl_tree.h", line 1934, in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol> > >::_M_lower_bound(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LFortran::Symbol> >*, std::_Rb_tree_node_base*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    if (!_M_impl._M_key_compare(_S_key(__x), __k))
  File "/usr/include/c++/9/bits/stl_function.h", line 386, in std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
    { return __x < __y; }
  File "/usr/include/c++/9/bits/basic_string.h", line 6232, in bool std::operator< <char, std::char_traits<char>, std::allocator<char> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    { return __lhs.compare(__rhs) < 0; }
  Binary file "/lib/x86_64-linux-gnu/libstdc++.so.6", in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::compare(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __nss_database_lookup()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in killpg()
Segfault: Signal SIGSEGV (segmentation fault) received
```
Could someone possibly try this PR and help with the bug, please?

**Comment by Shaikh-Ubaid** [comment]
The `wasm` to `x86` backend can be used as `lpython examples/expr2.py --backend wasm_x86`.

**Comment by Shaikh-Ubaid** [comment]
For https://github.com/lcompilers/lpython/pull/1223#issuecomment-1290921161, the `I am here` flow that is being generated using the `wasm` to `x86` backend is

```bash
(lp) ubaid@ubaid-Lenovo-ideapad-330-15ARR:~/OpenSource/lpython$ lpython examples/expr2.py --backend wasm_x86
I am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
I am here - 4
I am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
I am here - 4
I am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
I am here - 4
I am here - 1
I am here - 4
Call to imported function with index 5 not yet supportedI am here - 1
I am here - 2
I am here - 3
I am here - 4
I am here - 1
```

**Comment by Shaikh-Ubaid** [comment]
This https://github.com/lcompilers/lpython/pull/1223#issuecomment-1290921161 seems to be solved now.

**Comment by Shaikh-Ubaid** [comment]
The following example works with both the `x86` backends.

```python
def Main0():
    x: i32
    x = (2+3)*5
    y: i32
    z: i32
    y = 14
    z = 2
    print(x + y * z)

Main0()

# Not implemented yet in LPython:
#if __name__ == "__main__":
#    main()
```

```bash
(lp) ubaid@ubaid-Lenovo-ideapad-330-15ARR:~/OpenSource/lpython$ lpython examples/expr2.py --backend x86
(lp) ubaid@ubaid-Lenovo-ideapad-330-15ARR:~/OpenSource/lpython$ ./a.out 
53
(lp) ubaid@ubaid-Lenovo-ideapad-330-15ARR:~/OpenSource/lpython$ lpython examples/expr2.py --backend wasm_x86
(lp) ubaid@ubaid-Lenovo-ideapad-330-15ARR:~/OpenSource/lpython$ ./a.out 
53(lp) ubaid@ubaid-Lenovo-ideapad-330-15ARR:~/OpenSource/lpython$ 
```

**Comment by Shaikh-Ubaid** [comment]
**Notes:**
- We are needing to name the `main0` function as `Main0` so that it gets defined before the `_lpython_main_program()` which calls it.
- Only Non-Negative integers are currently supported as it seems `x86_assembler` does not support negative integers fully. (It seems there is no printing support for negative integers).
- Functions are supported, but returning values from functions is not yet supported.

**Comment by Shaikh-Ubaid** [comment]
This is ready. Please possibly review and please share feedback.

**Comment by certik** [comment]
At a high level this looks good. I need to review in more detail.

The changes in `lpython.cpp` seem long. We should try to figure out if there is an easier way to do this there, possibly in subsequent PRs.

**Comment by certik** [comment]
I will only have time next week to play with this --- in the meantime why don't you setup a CI to automatically test this? We'll need that anyway.

**Comment by Shaikh-Ubaid** [comment]
> in the meantime why don't you setup a CI to automatically test this? We'll need that anyway.

I was hoping to submit a PR for setting up the CI, but it seems the current `wasm` to `x86` backend is very early and does/might not support even basic integration tests. I am working on supporting return values from function calls. It seems to work. I will be sending a PR soon.

For CI, we might also need support for `visit_Assert()` as it seems several integration tests use `assert` in them.

**Review by certik — APPROVED** [review]

**Pull request merged**